### PR TITLE
Fix bug #6488: visit freesite looses the anchor

### PR DESCRIPTION
**Pull request by nextgens**
The better version of #551

### CODE REVIEW

File: src/freenet/clients/http/WelcomeToadlet.java [hunk 1]
```diff
@@ -265,6 +265,18 @@ public void handleMethodPOST(URI uri, HTTPRequest request, ToadletContext ctx) t
             writeHTMLReply(ctx, 200, "OK", pageNode.generate());
             request.freeParts();
             bucket.free();
+            return;
+        } else if (request.isPartSet("key")) {
+            if(!ctx.checkFormPassword(request)) return;
+            String key;
+            try {
+              key = URLDecoder.decode(new FreenetURI(request.getPartAsStringFailsafe("key", 256)).toURI("/").toString(), false);
```
Comment (bertm) [inline src/freenet/clients/http/WelcomeToadlet.java]: FreenetURI imposes a limit of 32767 (= Short.MAX_VALUE) to the key's length. Any reason to limit it to less than that?

File: src/freenet/clients/http/WelcomeToadlet.java [hunk 2]
```diff
@@ -265,6 +265,18 @@ public void handleMethodPOST(URI uri, HTTPRequest request, ToadletContext ctx) t
...
+            } catch (Exception e) {
+              sendErrorPage(ctx, "Invalid URI", "This URI is invalid!", e);
```
Comment (bertm) [inline src/freenet/clients/http/WelcomeToadlet.java]: This need l10n

Comment (nextgens) [inline src/freenet/clients/http/WelcomeToadlet.java]: I know that we've translated error messages in the past... but I'm not sure it's a good idea; do you insist on this?

File: src/freenet/clients/http/WelcomeToadlet.java (hunk 1 above)
Comment (nextgens) [inline src/freenet/clients/http/WelcomeToadlet.java]: I've copied/doubled what was above... but fair enough, we don't need a limit here

File: src/freenet/clients/http/WelcomeToadlet.java (hunk 2 above)
Comment (bertm) [inline src/freenet/clients/http/WelcomeToadlet.java]: I don't care much for the more obscure (internal) error messages, but this one can easily be triggered by the user mistyping stuff. This is user interface stuff after all.

I won't object to having this merged without being l10n'ed, but if you find the time, please do so.

File: src/freenet/clients/http/WelcomeToadlet.java [hunk 3]
```diff
@@ -265,6 +265,18 @@ public void handleMethodPOST(URI uri, HTTPRequest request, ToadletContext ctx) t
             writeHTMLReply(ctx, 200, "OK", pageNode.generate());
             request.freeParts();
             bucket.free();
+            return;
+        } else if (request.isPartSet("key")) {
+            if(!ctx.checkFormPassword(request)) return;
+            String key;
+            try {
+              key = URLDecoder.decode(new FreenetURI(request.getPartAsStringFailsafe("key", Short.MAX_VALUE)).toURI("/").toString(), false);
```
Comment (ArneBab) [inline src/freenet/clients/http/WelcomeToadlet.java]: could you split this line? (it’s longer than the visible width in Github)

**Pull request merged**